astroquery/cadc/core.py

```diff
@@ -141,9 +153,12 @@ def login(self, user=None, password=None, certificate_file=None):
         certificates.
         """
         # start with a new session
-        pyvo.dal.tap.s = requests.Session()
+        if not isinstance(self.cadctap._session, authsession.AuthSession):
+            raise TypeError('Cannot login with user provided session that is '
+                            'not an pyvo.autsession.AuthSession')
```

Comment:
missing an h in pyvo.autsession

Reply:
so it was my hope that people could also pass in other classes here, like just a normal requests session, but I doubt this will come up any time soon.

Reply:
Ah, I think I see, because you take the client certificate, you want it to be an authsession so you can call set_client_ceritificate, which only exists on that.  But for client certificates, it's not really a login anyway right?  Maybe just have client certificate use a different call than login?

I think another option might be just to set it to a fresh new authsession no matter what, after all, you're doing a login, so maybe that implies a logout of what was before?  Like in logout you're clearing everything (which I think is the right thing to do).

Anyway, nitpicking.

Reply:
`login` was here from the beginning (historical reasons). In fact it should be called `authenticate` and should allow multiple logins to the same service (cert, cookie etc).
I think I've tried it with a session of type `requests.Session` and it works.
Truth is I can probably login when the session is a simple `request.Session` which I'll change it to do. The difference is that such a simple session does not know to match credentials with the type of end point so that would be a reason for people to use 'AuthSession' instead.

Reply:
Updated a bunch more. Made it so it also uses user credentials with a `requests.Session`. Hopefully it's more clearer now. This is until we need to put it all together and have A&A working with other related services such as `Datalink` and `SODA`. It should be good for now.